zipkin-cassandra-core/src/main/java/org/twitter/zipkin/storage/cassandra/Repository.java

```diff
@@ -70,6 +70,23 @@
     private final PreparedStatement selectTraceDurationTail;
     private final Map<String,String> metadata;
 
+    private final ThreadLocal<Set<String>> writtenNames = new ThreadLocal<Set<String>>() {
+                private long hour = TimeUnit.MILLISECONDS.toHours(System.currentTimeMillis());
```

Comment:
> Different installations might use different TTL in Cassandra, in most cases at least 1 day, so cache TTL of 1hr is too fine.

We're not needing to match the index ttl value here.
For example if the cache ttl matched the index ttl, ie 3 days, then after 3 days all service and service-span names would disappear, despite there having been plenty of spans stored over the past 3 days with these values.

Describing it the other way around, with a cache ttl of one hour it means that the service and service-names will disappear after 2 days and 23 hours rather than the index ttl of 3 days.

So… the cache ttl needs to be some small proportion of the index ttl, and i'm thinking that a cache ttl of one hour is certainly big enough in regards to both performance (the reduction of writes) and to the UI (ensuring the service and spans names are available to choose from), and i don't think it's too small (is anyone setting the index ttl so low as just a few hours?)

> Second problem - the TTL is on the whole cache, not on individual entries.

This isn't a problem. This improvement is not trying to remove duplicate writes, simply improve performance (and the need for bucketing).


> Also, thread locals are unnecessary here, a concurrent map shared between threads would do.

concurrency has overhead, here it is a choice between letting the zipkin jvm handle it or letting cassandra handle it (ie duplicates writes). performance wise, which is the context for this improvement, cassandra is superior at handling the concurrency. hence the thread local approach.